File: src/main/java/org/jetbrains/StringJavaStreamBenchmark.java
package org.jetbrains;

import org.openjdk.jmh.annotations.*;
import org.openjdk.jmh.infra.Blackhole;

import java.util.ArrayList;
import java.util.concurrent.TimeUnit;
import java.util.stream.Stream;

@State(Scope.Thread)
@BenchmarkMode(Mode.Throughput)
@OutputTimeUnit(TimeUnit.SECONDS)
public class StringJavaStreamBenchmark {
    public static <E> ArrayList<E> makeList(Iterable<E> iter) {
        ArrayList<E> list = new ArrayList<E>();
        for (E item : iter) {
            list.add(item);
        }
        return list;
    }

    public ArrayList<String> data = makeList(JetbrainsPackage.stringValues());

    @Benchmark
    public Long filterAndCount() {
        return data.stream().filter(it -> it.length() % 2 == 0).count();
    }

    @Benchmark
    public void filterAndMap(final Blackhole bh) {
        Stream<String> stream = data.stream().filter(it -> it.length() % 2 == 0).map(it -> it + "x");
        stream.forEach(it -> bh.consume(it));
    }

    @Benchmark
    public Long countFiltered() {
        return data.stream().mapToLong(it -> it.length() % 2 == 0 ? 1 : 0).sum();
    }
}

File: src/main/kotlin/org/jetbrains/StringJavaStreamBenchmark.kt
package org.jetbrains

import org.openjdk.jmh.annotations.*
import java.util.concurrent.*
import java.util.stream.*
import org.openjdk.jmh.infra.Blackhole
import java.util.ArrayList

State(Scope.Thread)
BenchmarkMode(Mode.Throughput)
OutputTimeUnit(TimeUnit.SECONDS)
open class StringJavaStreamBenchmark {
    val list: ArrayList<String> = ArrayList(stringValues().toList())
    val getStreamMethod = list.javaClass.getMethod("stream")
    val data: Stream<String>
        get() = getStreamMethod.invoke(list) as Stream<String>

    Benchmark fun filterAndCount(bh: Blackhole) {
        val stream = data.filter(predicate<String>({ it.length % 2 == 0 }))!!
        bh.consume(stream.count())
    }

    Benchmark fun countFiltered(bh: Blackhole) {
        val stream = data.mapToLong(longFunction({ if (it.length % 2 == 0) 1L else 0L }))!!
        bh.consume(stream.sum())
    }

    Benchmark fun filterAndMap(bh: Blackhole) {
        val stream = data.filter(predicate<String>({ it.length % 2 == 0 }))!!.map(function<String, String>({ it + "x" }))!!
        stream.forEach(consume({ bh.consume(it) }))
    }
}
